### DISCUSSION TITLE
Store arbitrary data on the Context

### DISCUSSION THREAD

**Discussion by petermenkeepsilon** [TypedProject/ts-express-decorators#2159] [Q&A]
With Ts.ED v6, my team's application used a custom logger Layout that logged a message and added any values that were added to the Context with `ctx.set`. This worked because `Context` extended `Map`, so we could set values in a controller or service, and retrieve them with `ctx.forEach` in the Layout.

```ts
@Layout({ name: 'myLayout' })
export class MyLayout extends BaseLayout {
    public transform(loggingEvent: LogEvent, timezoneOffset?: number | undefined): string {
        const log: any = {
            categoryName: loggingEvent.categoryName,
            level: loggingEvent.level.toString(),
            message: loggingEvent.data.join(' '),
        };
        const store = PlatformAsyncHookContext?.getStore()?.getStore();
        if (store) {
            const correlationId = store.id;
            if (correlationId) {
                log.correlationId = correlationId;
            }

            // Check for any additional context needed to be added to the logs
            if (store.size) {
                log.data = {};
                store.forEach((value: any, key: any) => {
                    log.data[key] = value;
                });
            }
        }
        return JSON.stringify(log);
    }
}
```

With v7, `Context` no longer extends `Map` and instead it stores values in a private `#cache` variable. `ctx.set` still works, but there is no equivalent of `ctx.forEach` to retrieve all of the values in the `#cache`.

Is there a better way to store arbitrary data for the lifetime of a request so it can be used in the Layout?

We are currently looking into moving the data to the response locals or providing a service scoped to the request with `@Scope` to hold this data.

**Comment by Romakita** [comment]
Hello @petermenkeepsilon 

Just move your data into a sub map. 

```ts
$ctx.set(MY_KEYS, new Map())
$ctx.get<Map<string,any>(MY_KEYS).forEach()
```
> Note: Use foreach on ctx is not a good idea because Tsed an module store somes data in the map. Your log will have extra complex data.

You can also use a service context (Scope request is not necessary since `@InjectContext` exists)

```ts

@Injectable()
class MyContext {
  @InjectContext()
  $ctx: PlatformContext;

  $onRequest($initialCtx: PlatformContext) {
     $initialCtx.set(MyContext, new Map());
     // note $ctx.id can be used as correlationId (uuid)
   }

   get map(): Map<string, any>{
     return this.$ctx.get(MyContext);
   }
   // add any business method to control stored data in your request context.
   setProduct(p: Product){
      this.map.set(p);
   }
}


```

see you ;)

**Reply by petermenkeepsilon** [reply]
I'm trying the sub map. The Layout used to get the Context with `PlatformAsyncHookContext?.getStore()?.getStore()`, but I'm trying `@InjectContext()` now. However, it is always undefined. Is it possible to inject the Context (or any service) into a Layout?

**Reply by Romakita** [reply]
It’s always undefined if you are not under a request process. 

Also `@InjectContext` works only on injectable class. So you cannot use this decorator for your custom appender.

This feature works if you run your code under nodejs 16+ (or 14.13+).

To implement unit test you have to use a special util:

```ts

it("should run service with ctx",async () => {
    const $ctx = PlatformTest.getContextRequest();
    
    await runInContext($ctx, () => {
         return service.myMethod()
    })
})
```

See you

**Reply by Romakita** [reply]
Last point PlatformAsyncHook is deprecated in v6 and removed in v7. The code is directly implemented in ˋ@tsed/di` package.

see https://github.com/tsedio/tsed/blob/570277603a9e110ac1b5bfbf50cb23fc14f3cc0a/packages/di/src/utils/asyncHookContext.ts#L15

**Reply by petermenkeepsilon** [reply]
That makes sense, I'll try out getContext. Thank you!!
